sky/provision/kubernetes/network.py

```diff
@@ -0,0 +1,205 @@
+import os
+from typing import Any, Dict, List, Optional, Tuple
+
+from sky import skypilot_config
+from sky.provision.kubernetes import network_utils
+from sky.utils import ux_utils
+
+_PATH_PREFIX = "/skypilot/{cluster_name_on_cloud}/{port}"
+_LOADBALANCER_SERVICE_NAME = "{cluster_name_on_cloud}-skypilot-loadbalancer"
+
+
+def _get_port_mode() -> network_utils.KubernetesPortMode:
+    mode_str = skypilot_config.get_nested(
+        ('kubernetes', 'ports'),
+        network_utils.KubernetesPortMode.LOADBALANCER.value)
+    try:
+        port_mode = network_utils.KubernetesPortMode.from_str(mode_str)
+    except ValueError as e:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(str(e) + ' Please check: ~/.sky/config.yaml.') \
+                from None
+
+    return port_mode
+
+
+def open_ports(
+    cluster_name_on_cloud: str,
+    ports: List[str],
+    provider_config: Optional[Dict[str, Any]] = None,
+) -> None:
+    """See sky/provision/__init__.py"""
+    port_mode = _get_port_mode()
+
+    if port_mode == network_utils.KubernetesPortMode.LOADBALANCER:
+        _open_ports_using_loadbalancer(
+            cluster_name_on_cloud=cluster_name_on_cloud,
+            ports=ports,
+            provider_config=provider_config)
+    elif port_mode == network_utils.KubernetesPortMode.INGRESS:
+        _open_ports_using_ingress(cluster_name_on_cloud=cluster_name_on_cloud,
+                                  ports=ports,
+                                  provider_config=provider_config)
+
+
+def _open_ports_using_loadbalancer(
```

Comment:
I ran into a bug. Repro:

1. `sky launch -c jupyter jupyter.yaml`
```yaml
resources:
  ports: 8888

setup: pip install jupyter

run: jupyter lab --port 8888 --no-browser --ip=0.0.0.0
```

2. This works fine. Next, I run `sky cancel jupyter 1` to stop the server.
3. Then I update the YAML to use port 8889:
```yaml
resources:
  ports: 8889

setup: pip install jupyter

run: jupyter lab --port 8889 --no-browser --ip=0.0.0.0
```

4. When I `sky exec jupyter jupyter.yaml`, I get the following error:
```
D 11-02 16:16:24 command_runner.py:111] --- Proxy: ssh -tt -i ~/.ssh/sky-key -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -W %h:%p sky@127.0.0.1 -o ProxyCommand='/Users/romilb/.sky/port-forward-proxy-cmd.sh'  ---
D 11-02 16:16:30 cloud_vm_ray_backend.py:3013] Opening ports ['8888-8889'] for Kubernetes
D 11-02 16:16:30 skypilot_config.py:93] User config: kubernetes.ports -> loadbalancer
D 11-02 16:16:30 skypilot_config.py:157] Using config path: /Users/romilb/.sky/config.yaml
D 11-02 16:16:30 skypilot_config.py:160] Config loaded:
D 11-02 16:16:30 skypilot_config.py:160] {'kubernetes': {'ports': 'loadbalancer'}}
D 11-02 16:16:30 skypilot_config.py:166] Config syntax check passed.
Clusters
NAME     LAUNCHED     RESOURCES                                      STATUS  AUTOSTOP  COMMAND                       
jupyter  34 secs ago  1x Kubernetes(2CPU--2GB, ports=['8888-8889'])  INIT    -         sky launch -c jupyter tes...  

Traceback (most recent call last):
  File "/Users/romilb/tools/anaconda3/bin/sky", line 8, in <module>
    sys.exit(cli())
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 1163, in invoke
    return super().invoke(ctx)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 1439, in launch
    _launch_with_confirm(task,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/cli.py", line 845, in _launch_with_confirm
    sky.launch(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/execution.py", line 478, in launch
    _execute(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/execution.py", line 315, in _execute
    handle = backend.provision(task,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 332, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/utils/common_utils.py", line 311, in _record
    return f(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/backend.py", line 56, in provision
    return self._provision(task, to_provision, dryrun, stream_logs,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/cloud_vm_ray_backend.py", line 3006, in _provision
    self._update_after_cluster_provisioned(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/cloud_vm_ray_backend.py", line 3077, in _update_after_cluster_provisioned
    self._open_ports(handle)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/backends/cloud_vm_ray_backend.py", line 3017, in _open_ports
    provision_lib.open_ports(repr(cloud), handle.cluster_name_on_cloud,
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/__init__.py", line 41, in _wrapper
    return impl(*args, **kwargs)
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/kubernetes/network.py", line 35, in open_ports
    _open_ports_using_loadbalancer(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/kubernetes/network.py", line 60, in _open_ports_using_loadbalancer
    network_utils.create_or_replace_namespaced_loadbalancer(
  File "/Users/romilb/Romil/Berkeley/Research/sky-experiments/sky/provision/kubernetes/network_utils.py", line 97, in create_or_replace_namespaced_loadbalancer
    core_api.replace_namespaced_service(service_name, namespace, service_spec)
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/api/core_v1_api.py", line 28866, in replace_namespaced_service
    return self.replace_namespaced_service_with_http_info(name, namespace, body, **kwargs)  # noqa: E501
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/api/core_v1_api.py", line 28973, in replace_namespaced_service_with_http_info
    return self.api_client.call_api(
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/api_client.py", line 348, in call_api
    return self.__call_api(resource_path, method,
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/api_client.py", line 180, in __call_api
    response_data = self.request(
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/api_client.py", line 399, in request
    return self.rest_client.PUT(url,
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/rest.py", line 289, in PUT
    return self.request("PUT", url,
  File "/Users/romilb/tools/anaconda3/lib/python3.9/site-packages/kubernetes/client/rest.py", line 238, in request
    raise ApiException(http_resp=r)
kubernetes.client.exceptions.ApiException: (400)
Reason: Bad Request
HTTP response headers: HTTPHeaderDict({'Audit-Id': 'cbe1dd5e-96bf-46f4-9076-bf4f5acc57d3', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': 'cf43db27-7222-4741-adea-8b7f8f0e681e', 'X-Kubernetes-Pf-Prioritylevel-Uid': '17aff9c6-0fa7-4909-a811-4644194e9001', 'Date': 'Thu, 02 Nov 2023 23:16:30 GMT', 'Content-Length': '265'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Service in version \"v1\" cannot be handled as a Service: json: cannot unmarshal string into Go struct field ServicePort.spec.ports.port of type int32","reason":"BadRequest","code":400}
```

Reply:
Same seems to be the case when a port range is specified in the task yaml:
```
resources:
  ports: 8888-8891
```

Reply:
I will work on a fix for this.

Reply:
This is done in https://github.com/skypilot-org/skypilot/pull/2713/commits/4ae66029fba78e452aaf6ce9aec7313e91615c66